lib/application.js

```diff
@@ -398,85 +399,217 @@ app.enableAuth = function(options) {
+      // see https://github.com/strongloop/loopback/pull/3227
+      hasMultipleUserModelsConfig = !!(belongsToUser && belongsToUser.polymorphic);
     }
+
+    let modelSetup = {
+      model: model.modelName,
+      instanceOf: isInstanceOfUser ? 'User' : 'AccessToken',
+      hasMultipleUserModelsConfig,
+    };
+
+    // check if model config is consistent with already parsed models' config
+    if (self.hasMultipleUserModelsConfig === undefined) {
+      self.hasMultipleUserModelsConfig = hasMultipleUserModelsConfig;
+    } else if (self.hasMultipleUserModelsConfig !== hasMultipleUserModelsConfig) {
+      let error = new Error('User models setup is inconsistent');
+      error.modelSetup = modelSetup;
+      throw error;
+    }
+
+    // if no error, return the modelSetup for further logging purposes
+    return {modelSetup};
   }
 
   function verifyAccessTokenRelations(Model) {
+    if (!warnOnBadSetup) return;
     const belongsToUser = Model.relations && Model.relations.user;
     if (belongsToUser) return;
 
+    const modelFrom = Model.modelName;
     const relationsConfig = Model.settings.relations || {};
-    const userName = (relationsConfig.user || {}).model;
-    if (userName) {
-      console.warn(
-        'The model %j configures "belongsTo User-like models" relation ' +
-          'with target model %j. However, the model %j is not attached to ' +
-          'the application and therefore cannot be used by this relation. ' +
-          'This typically happens when the application has a custom ' +
-          'custom User subclass, but does not fix AccessToken relations ' +
-          'to use this new model.\n' +
-          'Learn more at http://ibm.biz/setup-loopback-auth',
-        Model.modelName, userName, userName);
+    const modelTo = (relationsConfig.user || {}).model;
+    if (modelTo) {
+      warnFn('CUSTOM_USER_MODEL_NOT_AVAILABLE', {modelFrom, modelTo});
       return;
     }
 
-    console.warn(
-      'The model %j does not have "belongsTo User-like model" relation ' +
-        'configured.\n' +
-        'Learn more at http://ibm.biz/setup-loopback-auth',
-      Model.modelName);
+    warnFn('MISSING_RELATION_TO_USER', {modelFrom});
   }
 
   function verifyUserRelations(Model) {
+    if (!warnOnBadSetup) return;
+
     const hasManyTokens = Model.relations && Model.relations.accessTokens;
     if (hasManyTokens) return;
 
+    const modelFrom = Model.modelName;
     const relationsConfig = Model.settings.relations || {};
-    const accessTokenName = (relationsConfig.accessTokens || {}).model;
-    if (accessTokenName) {
-      console.warn(
+    const modelTo = (relationsConfig.accessTokens || {}).model;
+    if (modelTo) {
+      warnFn('CUSTOM_ACCESS_TOKEN_MODEL_NOT_AVAILABLE', {modelFrom, modelTo});
+      return;
+    }
+
+    warnFn('MISSING_RELATION_TO_ACCESS_TOKEN', {modelFrom});
+  }
+};
+
+// function warnOnBadAuthModelRelations(code, args) {
+app._warnOnBadAuthModelRelations = function(code, args) {
```

Comment:
built-in warning function ("private")
i actually need it to be attached on the app object so i can swap it in the tests when calling `enableAuth()`